File: src/SIL.Machine.Shared/Translation/PhraseTranslationSuggester.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace SIL.Machine.Translation
{
	public class PhraseTranslationSuggester : ITranslationSuggester
	{
		public PhraseTranslationSuggester(double confidenceThreshold)
		{
			ConfidenceThreshold = confidenceThreshold;
		}

		public double ConfidenceThreshold { get; set; }

		public IEnumerable<int> GetSuggestedWordIndices(int prefixCount, bool isLastWordComplete,
			TranslationResult result)
		{
			int startingJ = prefixCount;
			if (!isLastWordComplete)
			{
				// if the prefix ends with a partial word and it has been completed,
				// then make sure it is included as a suggestion,
				// otherwise, don't return any suggestions
				if ((result.WordSources[startingJ - 1] & TranslationSources.Smt) != 0)
					startingJ--;
				else
					yield break;
			}

			int k = 0;
			while (k < result.Phrases.Count && result.Phrases[k].TargetSegmentCut <= startingJ)
				k++;

			for (; k < result.Phrases.Count; k++)
			{
				Phrase phrase = result.Phrases[k];
				if (phrase.Confidence >= ConfidenceThreshold)
				{
					for (int j = startingJ; j < phrase.TargetSegmentCut; j++)
					{
						string word = result.TargetSegment[j];
						TranslationSources sources = result.WordSources[j];
						if (sources == TranslationSources.None || word.All(char.IsPunctuation))
							yield break;
						yield return j;
					}
					startingJ = phrase.TargetSegmentCut;
				}
				else
				{
					yield break;
				}
			}
		}
	}
}

File: src/SIL.Machine.TS/src/translation/phrase-translation-suggester.ts
import XRegExp from 'xregexp';

import { TranslationResult } from './translation-result';
import { TranslationSources } from './translation-sources';
import { TranslationSuggester } from './translation-suggester';
import { TranslationSuggestion } from './translation-suggestion';

const ALL_PUNCT_REGEXP = XRegExp('^\\p{P}*$');

export class PhraseTranslationSuggester implements TranslationSuggester {
  confidenceThreshold: number = 0;

  getSuggestion(prefixCount: number, isLastWordComplete: boolean, result: TranslationResult): TranslationSuggestion {
    let startingJ = prefixCount;
    if (!isLastWordComplete) {
      // if the prefix ends with a partial word and it has been completed,
      // then make sure it is included as a suggestion,
      // otherwise, don't return any suggestions
      if ((result.wordSources[startingJ - 1] & TranslationSources.Smt) !== 0) {
        startingJ--;
      } else {
        return new TranslationSuggestion();
      }
    }

    let k = 0;
    while (k < result.phrases.length && result.phrases[k].targetSegmentCut <= startingJ) {
      k++;
    }

    let minConfidence = -1;
    const indices: number[] = [];
    for (; k < result.phrases.length; k++) {
      const phrase = result.phrases[k];
      if (phrase.confidence >= this.confidenceThreshold) {
        let hitBreakingWord = false;
        for (let j = startingJ; j < phrase.targetSegmentCut; j++) {
          const word = result.targetSegment[j];
          const sources = result.wordSources[j];
          if (sources === TranslationSources.None || ALL_PUNCT_REGEXP.test(word)) {
            hitBreakingWord = true;
            break;
          }
          indices.push(j);
        }
        if (minConfidence < 0 || phrase.confidence < minConfidence) {
          minConfidence = phrase.confidence;
        }
        startingJ = phrase.targetSegmentCut;
        if (hitBreakingWord) {
          break;
        }
      } else {
        break;
      }
    }

    return new TranslationSuggestion(indices, minConfidence < 0 ? 0 : minConfidence);
  }
}
